File: apps/ios/SignalASI/AgentClarificationPolicy.swift
import Foundation

enum AgentClarificationMode: String, Codable, CaseIterable, Identifiable {
  case execute = "EXECUTE"
  case askLocally = "ASK_LOCALLY"
  case askWithModel = "ASK_WITH_MODEL"

  var id: String { rawValue }
}

enum AgentClarificationQuestion: String, Codable, CaseIterable, Identifiable {
  case none = "NONE"
  case taskGoal = "TASK_GOAL"
  case codeOutcome = "CODE_OUTCOME"
  case controlAction = "CONTROL_ACTION"
  case researchTopic = "RESEARCH_TOPIC"
  case fileAction = "FILE_ACTION"
  case memoryContent = "MEMORY_CONTENT"
  case automationDetails = "AUTOMATION_DETAILS"

  var id: String { rawValue }
}

struct AgentClarificationDecision: Codable, Equatable {
  var mode: AgentClarificationMode
  var question: AgentClarificationQuestion

  init(
    mode: AgentClarificationMode,
    question: AgentClarificationQuestion = .none
  ) {
    self.mode = mode
    self.question = question
  }

  var shouldAsk: Bool {
    mode != .execute
  }
}

enum AgentClarificationPolicy {
  static func decide(
    goal: String,
    hasAttachments: Bool = false,
    hasConversationContext: Bool = false,
    preferenceMode: AgentPreferenceMode = .cautious
  ) -> AgentClarificationDecision {
    AgentPreferenceModePolicy.resolveClarification(
      mode: preferenceMode,
      goal: goal,
      baseline: decideBaseline(
        goal: goal,
        hasAttachments: hasAttachments,
        hasConversationContext: hasConversationContext
      )
    )
  }

  private static func decideBaseline(
    goal: String,
    hasAttachments: Bool,
    hasConversationContext: Bool
  ) -> AgentClarificationDecision {
    let normalized = normalize(goal)
    if normalized.isEmpty {
      if hasAttachments {
        return AgentClarificationDecision(mode: .askWithModel, question: .fileAction)
      }
      return ask(.taskGoal)
    }
    if hasConversationContext && isContextualFollowUp(normalized) {
      return execute
    }
    if hasAttachments && vagueRequests.contains(normalized) {
      return AgentClarificationDecision(mode: .askWithModel, question: .fileAction)
    }
    if vagueRequests.contains(normalized) {
      return hasConversationContext ? execute : ask(.taskGoal)
    }
    if isQuestion(normalized) || greetings.contains(normalized) {
      return execute
    }

    let missingQuestion: AgentClarificationQuestion?
    if codeRequestsWithoutOutcome.contains(normalized) {
      missingQuestion = .codeOutcome
    } else if controlRequestsWithoutAction.contains(normalized) {
      missingQuestion = .controlAction
    } else if researchRequestsWithoutTopic.contains(normalized) {
      missingQuestion = .researchTopic
    } else if fileRequestsWithoutAction.contains(normalized) {
      missingQuestion = .fileAction
    } else if memoryRequestsWithoutContent.contains(normalized) {
      missingQuestion = .memoryContent
    } else if automationRequestsWithoutDetails.contains(normalized) {
      missingQuestion = .automationDetails
    } else {
      missingQuestion = nil
    }
    if let missingQuestion = missingQuestion, !hasConversationContext {
      return ask(missingQuestion)
    }
    return execute
  }

  private static func ask(_ question: AgentClarificationQuestion) -> AgentClarificationDecision {
    AgentClarificationDecision(mode: .askLocally, question: question)
  }

  private static func normalize(_ value: String) -> String {
    let punctuationScalars: Set<UnicodeScalar> = [
      "\u{3002}", "\u{ff0c}", "\u{ff01}", "\u{ff1f}", "\u{ff1a}",
      "\u{ff1b}", "\u{201c}", "\u{201d}", "\u{2018}", "\u{2019}"
    ]
    let mapped = value.lowercased().unicodeScalars.map { scalar -> String in
      if CharacterSet.punctuationCharacters.contains(scalar) || punctuationScalars.contains(scalar) {
        return " "
      }
      return String(scalar)
    }.joined()
    return mapped
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func isQuestion(_ value: String) -> Bool {
    questionPrefixes.contains(where: value.hasPrefix) ||
      questionSuffixes.contains(where: value.hasSuffix)
  }

  private static func isContextualFollowUp(_ value: String) -> Bool {
    contextualFollowUps.contains(value) ||
      contextualReferences.contains(where: value.contains)
  }

  private static let execute = AgentClarificationDecision(mode: .execute)
  private static let greetings: Set<String> = [
    "hello", "hi", "hey", "good morning", "good afternoon", "good evening",
    "\u{4f60}\u{597d}", "\u{55e8}", "\u{65e9}\u{4e0a}\u{597d}",
    "\u{4e0b}\u{5348}\u{597d}", "\u{665a}\u{4e0a}\u{597d}"
  ]
  private static let questionPrefixes: Set<String> = [
    "what ", "why ", "how ", "when ", "where ", "which ", "who ",
    "can ", "could ", "would ", "is ", "are ", "do ", "does ",
    "\u{4ec0}\u{4e48}", "\u{4e3a}\u{4ec0}\u{4e48}", "\u{600e}\u{4e48}",
    "\u{5982}\u{4f55}", "\u{54ea}\u{4e2a}", "\u{54ea}\u{4e9b}",
    "\u{8c01}", "\u{80fd}\u{4e0d}\u{80fd}", "\u{53ef}\u{4ee5}"
  ]
  private static let questionSuffixes: Set<String> = [
    "\u{5417}", "\u{5462}", "\u{4e48}", "\u{600e}\u{4e48}\u{6837}",
    "\u{5982}\u{4f55}"
  ]
  private static let contextualFollowUps: Set<String> = [
    "continue", "go ahead", "do it", "try again", "retry", "keep going",
    "use this", "use that", "same as before", "make it better",
    "\u{7ee7}\u{7eed}", "\u{6267}\u{884c}", "\u{5c31}\u{8fd9}\u{6837}",
    "\u{6309}\u{8fd9}\u{4e2a}", "\u{518d}\u{8bd5}\u{8bd5}",
    "\u{91cd}\u{8bd5}", "\u{4fdd}\u{8bc1}\u{6b63}\u{786e}",
    "\u{7528}\u{8fd9}\u{4e2a}", "\u{548c}\u{4e4b}\u{524d}\u{4e00}\u{6837}",
    "\u{6309}\u{4e0a}\u{9762}\u{7684}\u{505a}"
  ]
  private static let contextualReferences: Set<String> = [
    " this", " that", " it", " above", " previous",
    "\u{8fd9}\u{4e2a}", "\u{90a3}\u{4e2a}", "\u{5b83}",
    "\u{4e0a}\u{9762}", "\u{4e4b}\u{524d}", "\u{521a}\u{624d}",
    "\u{524d}\u{9762}", "\u{8be5}\u{6587}\u{4ef6}", "\u{8fd9}\u{5f20}\u{56fe}"
  ]
  private static let vagueRequests: Set<String> = [
    "help me", "handle this", "do something", "take a look", "fix it",
    "improve it", "optimize it", "work on this", "please help",
    "\u{5e2e}\u{6211}", "\u{5e2e}\u{6211}\u{5f04}\u{4e00}\u{4e0b}",
    "\u{5904}\u{7406}\u{4e00}\u{4e0b}", "\u{5f04}\u{4e00}\u{4e0b}",
    "\u{770b}\u{770b}", "\u{5e2e}\u{6211}\u{770b}\u{770b}",
    "\u{4fee}\u{4e00}\u{4e0b}", "\u{4f18}\u{5316}\u{4e00}\u{4e0b}",
    "\u{6539}\u{8fdb}\u{4e00}\u{4e0b}", "\u{4f60}\u{770b}\u{7740}\u{529e}",
    "\u{7ed9}\u{6211}\u{7ed3}\u{679c}", "\u{5feb}\u{70b9}",
    "\u{4e0d}\u{884c}"
  ]
  private static let codeRequestsWithoutOutcome: Set<String> = [
    "write code", "write a program", "build an app", "create an app", "fix the code",
    "\u{5199}\u{4ee3}\u{7801}", "\u{5199}\u{4e2a}\u{7a0b}\u{5e8f}",
    "\u{5f00}\u{53d1}\u{4e00}\u{4e2a} app", "\u{505a}\u{4e00}\u{4e2a} app",
    "\u{4fee}\u{4ee3}\u{7801}"
  ]
  private static let controlRequestsWithoutAction: Set<String> = [
    "control my phone", "control the phone", "control my computer",
    "control the computer", "remote desktop",
    "\u{63a7}\u{5236}\u{624b}\u{673a}", "\u{64cd}\u{4f5c}\u{624b}\u{673a}",
    "\u{63a7}\u{5236}\u{7535}\u{8111}", "\u{64cd}\u{4f5c}\u{7535}\u{8111}",
    "\u{8fdc}\u{7a0b}\u{684c}\u{9762}"
  ]
  private static let researchRequestsWithoutTopic: Set<String> = [
    "research", "research this", "search", "search the web", "look it up",
    "\u{7814}\u{7a76}\u{4e00}\u{4e0b}", "\u{641c}\u{7d22}",
    "\u{641c}\u{4e00}\u{4e0b}", "\u{67e5}\u{4e00}\u{4e0b}",
    "\u{67e5}\u{8d44}\u{6599}"
  ]
  private static let fileRequestsWithoutAction: Set<String> = [
    "process the file", "handle the file", "work on the document",
    "\u{5904}\u{7406}\u{6587}\u{4ef6}", "\u{5904}\u{7406}\u{8fd9}\u{4e2a}\u{6587}\u{4ef6}",
    "\u{770b}\u{4e0b}\u{6587}\u{4ef6}"
  ]
  private static let memoryRequestsWithoutContent: Set<String> = [
    "remember this", "remember that", "save this to memory",
    "\u{8bb0}\u{4f4f}\u{8fd9}\u{4e2a}", "\u{8bb0}\u{4f4f}\u{8fd9}\u{4ef6}\u{4e8b}",
    "\u{5b58}\u{5230}\u{8bb0}\u{5fc6}"
  ]
  private static let automationRequestsWithoutDetails: Set<String> = [
    "create an automation", "make a workflow", "schedule a task", "remind me",
    "\u{521b}\u{5efa}\u{81ea}\u{52a8}\u{5316}", "\u{5efa}\u{4e00}\u{4e2a}\u{5de5}\u{4f5c}\u{6d41}",
    "\u{8bbe}\u{7f6e}\u{5b9a}\u{65f6}\u{4efb}\u{52a1}", "\u{63d0}\u{9192}\u{6211}"
  ]
}

File: apps/android/app/src/main/java/com/signalasi/chat/AgentClarificationPolicy.kt
package com.signalasi.chat

import java.util.Locale

enum class AgentClarificationMode {
    EXECUTE,
    ASK_LOCALLY,
    ASK_WITH_MODEL
}

enum class AgentClarificationQuestion {
    NONE,
    TASK_GOAL,
    CODE_OUTCOME,
    CONTROL_ACTION,
    RESEARCH_TOPIC,
    FILE_ACTION,
    MEMORY_CONTENT,
    AUTOMATION_DETAILS
}

data class AgentClarificationDecision(
    val mode: AgentClarificationMode,
    val question: AgentClarificationQuestion = AgentClarificationQuestion.NONE
) {
    val shouldAsk: Boolean
        get() = mode != AgentClarificationMode.EXECUTE
}

object AgentClarificationPolicy {
    fun decide(
        goal: String,
        hasAttachments: Boolean = false,
        hasConversationContext: Boolean = false,
        preferenceMode: AgentPreferenceMode = AgentPreferenceMode.CAUTIOUS
    ): AgentClarificationDecision = AgentPreferenceModePolicy.resolveClarification(
        mode = preferenceMode,
        goal = goal,
        baseline = decideBaseline(goal, hasAttachments, hasConversationContext)
    )

    private fun decideBaseline(
        goal: String,
        hasAttachments: Boolean,
        hasConversationContext: Boolean
    ): AgentClarificationDecision {
        val normalized = normalize(goal)
        if (normalized.isBlank()) {
            return if (hasAttachments) EXECUTE else ask(AgentClarificationQuestion.TASK_GOAL)
        }

        if (hasConversationContext && isContextualFollowUp(normalized)) {
            return EXECUTE
        }
        if (hasAttachments && normalized in VAGUE_REQUESTS) {
            return EXECUTE
        }
        if (normalized in VAGUE_REQUESTS) {
            return if (hasConversationContext) EXECUTE else ask(AgentClarificationQuestion.TASK_GOAL)
        }
        if (isQuestion(normalized) || isGreeting(normalized)) return EXECUTE

        val missingQuestion = when {
            normalized in CODE_REQUESTS_WITHOUT_OUTCOME ->
                AgentClarificationQuestion.CODE_OUTCOME
            normalized in CONTROL_REQUESTS_WITHOUT_ACTION ->
                AgentClarificationQuestion.CONTROL_ACTION
            normalized in RESEARCH_REQUESTS_WITHOUT_TOPIC ->
                AgentClarificationQuestion.RESEARCH_TOPIC
            normalized in FILE_REQUESTS_WITHOUT_ACTION ->
                AgentClarificationQuestion.FILE_ACTION
            normalized in MEMORY_REQUESTS_WITHOUT_CONTENT ->
                AgentClarificationQuestion.MEMORY_CONTENT
            normalized in AUTOMATION_REQUESTS_WITHOUT_DETAILS ->
                AgentClarificationQuestion.AUTOMATION_DETAILS
            else -> null
        }
        if (missingQuestion != null && !hasConversationContext) {
            return ask(missingQuestion)
        }
        return EXECUTE
    }

    private fun ask(question: AgentClarificationQuestion) =
        AgentClarificationDecision(AgentClarificationMode.ASK_LOCALLY, question)

    private fun normalize(value: String): String = value
        .lowercase(Locale.US)
        .replace(Regex("[\\p{Punct}\u3002\uff0c\uff01\uff1f\uff1a\uff1b\u201c\u201d\u2018\u2019]+"), " ")
        .replace(Regex("\\s+"), " ")
        .trim()

    private fun isQuestion(value: String): Boolean =
        value.startsWith(QUESTION_PREFIXES) ||
            value.endsWith(QUESTION_SUFFIXES)

    private fun isGreeting(value: String): Boolean = value in GREETINGS

    private fun isContextualFollowUp(value: String): Boolean =
        value in CONTEXTUAL_FOLLOW_UPS ||
            CONTEXTUAL_REFERENCES.any(value::contains)

    private fun String.startsWith(prefixes: Set<String>): Boolean =
        prefixes.any(::startsWith)

    private fun String.endsWith(suffixes: Set<String>): Boolean =
        suffixes.any(::endsWith)

    private val EXECUTE = AgentClarificationDecision(AgentClarificationMode.EXECUTE)

    private val GREETINGS = setOf(
        "hello", "hi", "hey", "good morning", "good afternoon", "good evening",
        "\u4f60\u597d", "\u55e8", "\u65e9\u4e0a\u597d", "\u4e0b\u5348\u597d", "\u665a\u4e0a\u597d"
    )
    private val QUESTION_PREFIXES = setOf(
        "what ", "why ", "how ", "when ", "where ", "which ", "who ",
        "can ", "could ", "would ", "is ", "are ", "do ", "does ",
        "\u4ec0\u4e48", "\u4e3a\u4ec0\u4e48", "\u600e\u4e48", "\u5982\u4f55",
        "\u54ea\u4e2a", "\u54ea\u4e9b", "\u8c01", "\u80fd\u4e0d\u80fd", "\u53ef\u4ee5"
    )
    private val QUESTION_SUFFIXES = setOf(
        "\u5417", "\u5462", "\u4e48", "\u600e\u4e48\u6837", "\u5982\u4f55"
    )
    private val CONTEXTUAL_FOLLOW_UPS = setOf(
        "continue", "go ahead", "do it", "try again", "retry", "keep going",
        "use this", "use that", "same as before", "make it better",
        "\u7ee7\u7eed", "\u6267\u884c", "\u5c31\u8fd9\u6837", "\u6309\u8fd9\u4e2a",
        "\u518d\u8bd5\u8bd5", "\u91cd\u8bd5", "\u4fdd\u8bc1\u6b63\u786e", "\u7528\u8fd9\u4e2a",
        "\u548c\u4e4b\u524d\u4e00\u6837", "\u6309\u4e0a\u9762\u7684\u505a"
    )
    private val CONTEXTUAL_REFERENCES = setOf(
        " this", " that", " it", " above", " previous",
        "\u8fd9\u4e2a", "\u90a3\u4e2a", "\u5b83", "\u4e0a\u9762", "\u4e4b\u524d",
        "\u521a\u624d", "\u524d\u9762", "\u8be5\u6587\u4ef6", "\u8fd9\u5f20\u56fe"
    )
    private val VAGUE_REQUESTS = setOf(
        "help me", "handle this", "do something", "take a look", "fix it",
        "improve it", "optimize it", "work on this", "please help",
        "\u5e2e\u6211", "\u5e2e\u6211\u5f04\u4e00\u4e0b", "\u5904\u7406\u4e00\u4e0b",
        "\u5f04\u4e00\u4e0b", "\u770b\u770b", "\u5e2e\u6211\u770b\u770b", "\u4fee\u4e00\u4e0b",
        "\u4f18\u5316\u4e00\u4e0b", "\u6539\u8fdb\u4e00\u4e0b", "\u4f60\u770b\u7740\u529e",
        "\u7ed9\u6211\u7ed3\u679c", "\u5feb\u70b9", "\u4e0d\u884c"
    )
    private val CODE_REQUESTS_WITHOUT_OUTCOME = setOf(
        "write code", "write a program", "build an app", "create an app", "fix the code",
        "\u5199\u4ee3\u7801", "\u5199\u4e2a\u7a0b\u5e8f", "\u5f00\u53d1\u4e00\u4e2a app",
        "\u505a\u4e00\u4e2a app", "\u4fee\u4ee3\u7801"
    )
    private val CONTROL_REQUESTS_WITHOUT_ACTION = setOf(
        "control my phone", "control the phone", "control my computer",
        "control the computer", "remote desktop",
        "\u63a7\u5236\u624b\u673a", "\u64cd\u4f5c\u624b\u673a",
        "\u63a7\u5236\u7535\u8111", "\u64cd\u4f5c\u7535\u8111", "\u8fdc\u7a0b\u684c\u9762"
    )
    private val RESEARCH_REQUESTS_WITHOUT_TOPIC = setOf(
        "research", "research this", "search", "search the web", "look it up",
        "\u7814\u7a76\u4e00\u4e0b", "\u641c\u7d22", "\u641c\u4e00\u4e0b",
        "\u67e5\u4e00\u4e0b", "\u67e5\u8d44\u6599"
    )
    private val FILE_REQUESTS_WITHOUT_ACTION = setOf(
        "process the file", "handle the file", "work on the document",
        "\u5904\u7406\u6587\u4ef6", "\u5904\u7406\u8fd9\u4e2a\u6587\u4ef6", "\u770b\u4e0b\u6587\u4ef6"
    )
    private val MEMORY_REQUESTS_WITHOUT_CONTENT = setOf(
        "remember this", "remember that", "save this to memory",
        "\u8bb0\u4f4f\u8fd9\u4e2a", "\u8bb0\u4f4f\u8fd9\u4ef6\u4e8b", "\u5b58\u5230\u8bb0\u5fc6"
    )
    private val AUTOMATION_REQUESTS_WITHOUT_DETAILS = setOf(
        "create an automation", "make a workflow", "schedule a task", "remind me",
        "\u521b\u5efa\u81ea\u52a8\u5316", "\u5efa\u4e00\u4e2a\u5de5\u4f5c\u6d41",
        "\u8bbe\u7f6e\u5b9a\u65f6\u4efb\u52a1", "\u63d0\u9192\u6211"
    )
}
